A67-xds-dynamic-parameters.md

```diff
@@ -0,0 +1,246 @@
+    config contains settings for each authority.
+-   We have a shared pool of xDS channel objects, so if multiple authorities use
+    the same xDS server, they will share the same xDS channel object.
+-   The code that constructs the ADS request already knows the set of resource
+    names to send, and that list may include resources from multiple authorities
+    that are sharing the same xDS server.
+
+The changes this design proposes are:
+
+-   The bootstrap config will contain the dynamic parameters to use for each
+    authority.
+-   The code to construct the ADS request will be changed such that for each
+    authority for which there are resource names in the request, it looks up the
+    dynamic parameters for that authority in the bootstrap config. If the
+    dynamic parameters are not empty, that will trigger using the
+    `resource_locators` field instead of the `resource_names` field in the
+    request
+-   The code to parse the ADS response will be changed to look in the
+    `resource_name` field and use this instead of the name field if it exists.
+    As this `XdsClient` is only a leaf client, and not a caching proxy, we can
+    ignore the dynamic_parameter_constraints field of the ResourceName message.
+
+The updated code to construct the ADS request would look something like this
+(pseudo code):
+
+```text
+// Example authority_resource_names:
+[
+  // Resources for authority "foo.example.com"
+  { resource_names: ["bar", "baz"], dynamic_parameters: {"env": "prod"} },
+
+  // Resources for authority "zzz.example.com"
+  { resource_names: ["yyy"], dynamic_parameters: {"env": "prod", "version": "v2"} },
+]
+
+function ConstructADSRequest(authority_resource_names)
+  ...
+  foreach resource_names, dynamic_parameters in authority_resource_names
+    if dynamic_parameters is empty
+      request.resource_names = resource_names
+    else
+      foreach name in resource_names
+        request.resource_locators.add(name, dynamic_parameters)
+  ...
+```
+
+### Temporary environment variable protection
+
+During initial development, the GRPC_EXPERIMENTAL_XDS_DYNAMIC_PARAMETERS
+environment variable will control whether ADS requests will use the
+`resource_locators` field with dynamic parameters populated, instead of the
+`resource_names` field. This will help guard against if an xDS server does not
+yet support dynamic parameters, but the
+[bootstrap generator](https://github.com/GoogleCloudPlatform/traffic-director-grpc-bootstrap)
+already sets the `dynamic_parameters` fields.
+
+## Rationale
+
+The following sub-sections describe tradeoffs that were made in this design.
+
+### No changas to dynamic parameters after startup
+
+For our initial support of dynamic parameters, we can let the parameter
+key-value pairs be only read at startup. A client restart will be required when
+changing the parameters.
+
+### No metadata reuse
+
+An idea in TP2 is to reuse the node metadata field as dynamic parameters. I
```

Comment:
I think there are a few key points to call out here:
- We need an explicit knob in the bootstrap config that tells us whether or not to use dynamic parameters, and using the presence of the `dynamic_parameters` field gives us that.  If we tried to automatically convert node metadata into dynamic parameters, we'd need a separate field to control whether dynamic parameters should be used.
- The suggestion in TP2 was really intended as a possible migration path for Envoy, where basically everyone needs to use node metadata because of the way that Envoy does wildcard LDS and CDS subscriptions, where the server uses node metadata to decide what resources to send.  In contrast, node metadata is not as important in the gRPC xDS ecosystem, so this migration path is not considered important.
- Converting node metadata to dynamic parameters would be somewhat complex to implement, since node metadata can have structured values, whereas dynamic parameter values are always just strings.

Reply:
Rewrote this section to include these key points.